packages/coreutils/src/url.ts

```diff
@@ -156,6 +156,153 @@ export namespace URLExt {
+   * Format: data:[<mediatype>][;base64],<data>
+   * Default MIME type is "text/plain;charset=US-ASCII" per the specification.
+   */
+  export function parseDataURI(dataURI: string): {
+    mimeType: string;
+    isBase64: boolean;
+    data: string;
+  } | null {
+    try {
+      // Verify it has the data: protocol
+      if (!dataURI.startsWith('data:')) {
+        return null;
+      }
+
+      // Find the comma that separates metadata from data
+      const commaIndex = dataURI.indexOf(',');
+      if (commaIndex === -1) {
+        return null;
+      }
+
+      // Extract metadata (everything between 'data:' and ',')
+      const metadata = dataURI.slice(5, commaIndex);
+      const data = dataURI.slice(commaIndex + 1);
+
+      // Check if data is base64-encoded
+      const isBase64 = metadata.endsWith(';base64');
+
+      // Extract MIME type (remove ;base64 suffix if present)
+      let mimeType = isBase64 ? metadata.slice(0, -7) : metadata;
+
+      // Default MIME type per RFC 2397
+      if (!mimeType) {
+        mimeType = 'text/plain;charset=US-ASCII';
+      }
+
+      return { mimeType, isBase64, data };
+    } catch (error) {
+      console.error('Error parsing data URI:', error);
+      return null;
+    }
+  }
+
+  /**
+   * Parse a text/uri-list string and return the first URI.
+   *
+   * @param uriList - The text/uri-list formatted string
+   *
+   * @returns The first URI found, or null if none
+   *
+   * #### Notes
+   * Parses text/uri-list format per RFC 2483.
+   * - Lines are terminated with CRLF pairs
+   * - Lines starting with '#' are comments and are ignored
+   * - Returns the first valid URI found without parsing the entire list
+   *
+   * See https://www.iana.org/assignments/media-types/text/uri-list
+   */
+  export function parseUriListFirst(uriList: string): string | null {
+    if (!uriList) {
+      return null;
+    }
+
+    let start = 0;
+    while (start < uriList.length) {
+      // Find the next CRLF
+      let end = uriList.indexOf('\r\n', start);
+      if (end === -1) {
+        // No more CRLF, take the rest of the string
+        end = uriList.length;
+      }
+
+      // Skip comments (lines starting with #)
+      if (uriList[start] !== '#') {
+        return uriList.slice(start, end);
+      }
+
+      // Move past the CRLF for next iteration
+      start = end + 2;
+    }
+
+    return null;
+  }
+
+  /**
+   * Convert a data URI to a Blob.
+   *
+   * @param dataURI - The data URI to convert (e.g., "data:image/png;base64,iVBORw0KG...")
+   *
+   * @returns A Blob object or null if conversion fails
+   *
+   * #### Notes
+   * This function handles both base64-encoded and percent-encoded data URIs.
+   * - Base64 data: Uses modern Uint8Array.fromBase64() if available, falls back to atob()
+   * - Non-base64 data: Percent-decoded (reserved characters per RFC 3986)
+   */
+  export function dataURItoBlob(dataURI: string): Blob | null {
```

Comment:
https://github.com/jupyterlab/jupyterlab/pull/17767 is also adding a function like that